neutron/agent/dhcp_agent.py

```diff
@@ -129,6 +130,53 @@ def run(self):
         self.periodic_resync()
 
     def call_driver(self, action, network, **action_kwargs):
+        # Spawn a thread for this network, if there isn't one already.
+        if network.id not in self.driver_queues:
+            LOG.info("Create queue and thread for network %s", network)
+            self.driver_queues[network.id] = eventlet.queue.Queue()
+            eventlet.spawn(self._call_driver_thread,
+                           network,
```

Comment:
I think network is mutable so it's not safe to pass it to the other thread like this, that's why I moved all the access to the cache onto my worker thread.  It removed all the possibility of multi-threaded access to the cache.

You could fix that up by taking a deep copy of the network to send to the other thread, but that might be expensive if there are 10k ports in there.  Alternatively, you could lock around the cache access on the worker thread but that's more complex than my solution so I don't think it's a good plan!